samcli/local/lambdafn/runtime.py

```diff
@@ -161,28 +242,278 @@ def _get_code_dir(self, code_path):
+        return container
+
+    def _on_invoke_done(self, container):
+        """
+        Cleanup the created resources, just before the invoke function ends.
+        In warm containers, the running containers will be closed just before the end of te command execution,
+        so no action is done here
 
-            elif os.path.isfile(code_path) and code_path.endswith(self.SUPPORTED_ARCHIVE_EXTENSIONS):
+        Parameters
+        ----------
+        container: Container
+           The current running container
+        """
+
+    def _add_function_to_observer(self, function_config):
+        """
+        observe the function code path, so the function container be reload if its code got changed.
 
-                decompressed_dir = _unzip_file(code_path)
-                yield decompressed_dir
+        Parameters
+        ----------
+        function_config: FunctionConfig
+            Configuration of the function to create a new Container for it.
+        """
+        with self._lock:
+            code_paths = [function_config.code_abs_path]
+            if function_config.layers:
+                code_paths += [layer.codeuri for layer in function_config.layers]
+            # get all sub directories to be observed, so if the any of the sub directories got changed
+            # we can map it to the correct lambda function
+            code_paths_sub_folders = []
+            for code_path in code_paths:
+                code_paths_sub_folders += [code_path]
+                if not os.path.isfile(code_path):
+                    code_paths_sub_folders += [sub_dir[0] for sub_dir in os.walk(code_path)]
+
+            # observe all the code directories && files
+            for code_path in code_paths_sub_folders:
+                if os.path.isfile(code_path):
+                    checksum = file_checksum(code_path)
+                else:
+                    # observe the build directory instead of the lambda function directory in case of runtime languages
+                    # that needs compilation
+                    if ".aws-sam" in code_path:
+                        code_path = os.path.abspath(os.path.join(code_path, os.pardir))
+                    checksum = dir_checksum(code_path)
+                self._observed_paths_checksum[code_path] = checksum
+
+                functions = self._observed_paths.get(code_path, [])
+                functions += [function_config.name]
+                self._observed_paths[code_path] = functions
+                self._observer.schedule(self._code_change_handler, code_path, recursive=True)
+            function_paths = self._function_paths.get(function_config.name, [])
+            function_paths += code_paths_sub_folders
+            self._function_paths[function_config.name] = function_paths
+            if not self._observer.is_alive():
+                self._observer.start()
 
+    def _configure_interrupt(self, function_name, timeout, container, is_debugging):
```

Comment:
Looks like this function is not using any resources in this class, and it is handling interrupt on running docker image. It will be better if we move this into a utility method with a generic signature, so that it can be used with something else as well.

Reply:
Ohh I see, thanks for explaining!